File: 3816. Lexicographically Smallest String After Deleting Duplicate Characters.cpp
class Solution {
public:
    string lexSmallestAfterDeletion(string s) {
        int freq[128] = {};
        for (int i = 0; i < s.length(); i++) {
            freq[s[i]]++;
        }

        stack<char> st;
        for (int i = 0; i < s.length(); i++) {
            while (st.size() > 0 && st.top() > s[i] && freq[st.top()] > 1) {
                freq[st.top()]--;
                st.pop();
            }
            st.push(s[i]); 
        }

        char dup = st.top();
        while (st.size() > 0 && freq[st.top()] > 1) {
            freq[st.top()]--;
            st.pop();
        }

        string sb;
        while (!st.empty()) {
            sb += st.top();
            st.pop();
        }

        reverse(sb.begin(), sb.end());

        return sb;
    }
};

File: 3816. Lexicographically Smallest String After Deleting Duplicate Characters.java
class Solution {
    public String lexSmallestAfterDeletion(String s) {
        int[] freq = new int[128];
        for(int i = 0; i < s.length(); i++){
            freq[s.charAt(i)]++;
        }

        Stack<Character> st = new Stack<>();
        for(int i = 0; i < s.length(); i++){
            while(st.size() > 0 && st.peek() > s.charAt(i) && freq[st.peek()] > 1){
                freq[st.peek()]--;
                st.pop();
            }
            st.push(s.charAt(i));
        }

        char dup = st.peek();
        while(st.size() > 0 && freq[st.peek()] > 1){
            freq[st.peek()]--;
            st.pop();
        }

        StringBuilder sb = new StringBuilder();
        while(!st.isEmpty()) sb.append(st.pop());

        return sb.reverse().toString(); 
    }
}